File: programmers/SummerWinter Coding(~2018)/방문 길이(Level3).py
'''
    Code written on Nov 3, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    Summer/Winter Coding(~2018): 방문 길이(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/49994
'''
def solution(dirs):
    command = {'U': (0, 1), 'D': (0, -1), 'R': (1, 0), 'L': (-1, 0)}
    x_coor, y_coor = 0, 0
    ways = set()
    for direction in dirs:
        movement_x, movement_y = command[direction]
        next_x_coor = x_coor + movement_x
        next_y_coor = y_coor + movement_y
        if (-5 <= next_x_coor <= 5 and -5 <= next_y_coor <= 5):
            ways.add((x_coor, y_coor, next_x_coor, next_y_coor))
            ways.add((next_x_coor, next_y_coor, x_coor, y_coor))
            x_coor, y_coor = next_x_coor, next_y_coor
    return len(ways)//2

File: programmers/SummerWinter Coding(~2018)/방문 길이(Level3).java
/*
    Code written on Dec 8, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    Summer/Winter Coding(~2018): 방문 길이(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/49994
*/
import java.util.HashSet;

class Solution {
    public int solution(String dirs) {
        HashSet<String> pastRoad = new HashSet<>();
        int[] coord = new int[2];
        int[] newCoord;

        for (char dir: dirs.toCharArray()) {
            newCoord = moveCoord(dir, coord);
            if ((newCoord[0] <= 5) && (newCoord[0] >= -5) &&
                    (newCoord[1] <= 5) && (newCoord[1] >= -5)) {
                pastRoad.add(String.format("%d%d%d%d", 
                        coord[0], coord[1], newCoord[0], newCoord[1]));
                pastRoad.add(String.format("%d%d%d%d", 
                        newCoord[0], newCoord[1], coord[0], coord[1]));
                coord = newCoord;
            }
        }
        return pastRoad.size() / 2;
    }

    public int[] moveCoord(char dir, int[] coord) {
        int[] newCoord = coord.clone();
        
        switch (dir) {
            case 'U':
                newCoord[1] = coord[1] + 1;
                break;
            case 'D':
                newCoord[1] = coord[1] - 1;
                break;
            case 'L':
                newCoord[0] = coord[0] - 1;
                break;
            case 'R':
                newCoord[0] = coord[0] + 1;
                break;
        }
        return newCoord;
    }
}
